### PR TITLE
Fix some overlay algorithms output multipoint geometries but output layer is single point, causing insertion errors (3.22)

### LINKED ISSUE #49456: Native CLIP function
### What is the bug or the crash?

With QGIS  version over 3.16 (e.g. 3.22.5 or 3.26)

When clipping vector points layer with another vector layer (e.g. polygon vector layer) and exporting the result in *.shp in a folder, the log reports a bug/crash: (Erreur OGR : Attempt to write non-point (MULTIPOINT) geometry to point shapefile).

Bug not reproduced with temporrary layer as  output or with other formats as output (e.g geojson, and so on).

### Steps to reproduce the issue

1. Load two vector layers. One with points and the other with a polygon.
2.  Use the native function for clipping (CLIP).
3.  Select points as first layer and polygno as the overlapping one.
4. Define the output for clipping as a *.shp in a folder.
5. Run the native function.
6. The log reports (Erreur OGR : Attempt to write non-point (MULTIPOINT) geometry to point shapefile).

### Versions

3.22.5 or 3.26 (maybe all version over 3.16.11)

### Supported QGIS version

- [x] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by nyalldawson**
Manual backport of https://github.com/qgis/QGIS/pull/50567

Fixes #49456

(cherry picked from commit c9a2a5efa8ec9a0430ec503823649f1ac4831ed0)

### CODE REVIEW

File: src/analysis/processing/qgsalgorithmdifference.cpp [hunk 1]
```diff
@@ -108,7 +108,7 @@ QVariantMap QgsDifferenceAlgorithm::processAlgorithm( const QVariantMap &paramet
 
   long count = 0;
   const long total = sourceA->featureCount();
-  QgsOverlayUtils::difference( *sourceA, *sourceB, *sink, context, feedback, count, total, QgsOverlayUtils::OutputA );
+  QgsOverlayUtils::difference( *sourceA, *sourceB, *sink, context, feedback, count, total, QgsOverlayUtils::OutputA, QgsOverlayUtils::SanitizeFlag::DontPromotePointGeometryToMultiPoint ); );
```
Comment (agiudiceandrea) [inline src/analysis/processing/qgsalgorithmdifference.cpp:111]:
```suggestion
  QgsOverlayUtils::difference( *sourceA, *sourceB, *sink, context, feedback, count, total, QgsOverlayUtils::OutputA, QgsOverlayUtils::SanitizeFlag::DontPromotePointGeometryToMultiPoint );
```

**Comment by agiudiceandrea** [comment]
Hi @nyalldawson, what happened to this backport? It seem's it was not in time for 3.12.12, nor for 3.22.13.

File: src/analysis/processing/qgsalgorithmdifference.cpp (hunk 1 above)
Comment (agiudiceandrea) [inline src/analysis/processing/qgsalgorithmdifference.cpp:111]:
```suggestion
  QgsOverlayUtils::difference( *sourceA, *sourceB, *sink, context, feedback, count, total, QgsOverlayUtils::OutputA, QgsOverlayUtils::SanitizeFlag::DontPromotePointGeometryToMultiPoint );
```

**Review by nirvn — APPROVED** [review]

**Comment by ShebMichel** [comment]
Is this issue fixed? 
QGIS version
3.26.2-Buenos Aires
QGIS code revision
feec3d3b12f
Qt version
5.15.3
Python version
3.9.5
GDAL/OGR version
3.5.1
PROJ version
9.0.1
EPSG Registry database version
v10.064 (2022-05-19)
GEOS version
3.10.3-CAPI-1.16.1
SQLite version
3.38.1
PDAL version
2.4.2
PostgreSQL client version
unknown
SpatiaLite version
5.0.1
QWT version
6.1.6
QScintilla2 version
2.13.1
OS version
Windows 10 Version 2009
Active Python plugins
canvas_coordinates
1.0
ClipMultipleLayers
3.2.0
create_layer_from_selected_features
1.2
finddataset
1.1.0
firstaid
2.2.1
LayerStyleLoader
0.2
mask
1.10.2
pg_raster_import
3.1.0
pluginbuilder3
3.2.1
plugin_reloader
0.9.3
db_manager
0.1.20
MetaSearch
0.3.6
processing
2.12.99
sagaprovider
2.12.99
loop_plugin
0.3

**Comment by ShebMichel** [comment]
Hi I am having the same issue. basically, I cannot clip point layer using roi created from polygone layer. Below is my qgis specs:

QGIS version
3.26.2-Buenos Aires
QGIS code revision
feec3d3b12f
Qt version
5.15.3
Python version
3.9.5
GDAL/OGR version
3.5.1
PROJ version
9.0.1
EPSG Registry database version
v10.064 (2022-05-19)
GEOS version
3.10.3-CAPI-1.16.1
SQLite version
3.38.1
PDAL version
2.4.2
PostgreSQL client version
unknown
SpatiaLite version
5.0.1
QWT version
6.1.6
QScintilla2 version
2.13.1
OS version
Windows 10 Version 2009




Active Python plugins
canvas_coordinates
1.0
ClipMultipleLayers
3.2.0
create_layer_from_selected_features
1.2
finddataset
1.1.0
firstaid
2.2.1
LayerStyleLoader
0.2
mask
1.10.2
pg_raster_import
3.1.0
pluginbuilder3
3.2.1
plugin_reloader
0.9.3
db_manager
0.1.20
MetaSearch
0.3.6
processing
2.12.99
sagaprovider
2.12.99
loop_plugin
0.3
Thanks

**Comment by agiudiceandrea** [comment]
@ShebMichel you are uing and old no longer supported QGIS version. Please use a currently supported QGIS version.

**Pull request merged**